warp-core-macros/src/main/scala/com/workday/warp/monadic/WarpMacros.scala

```diff
@@ -0,0 +1,149 @@
+
+  def deriveTestIdsImpl[T: ctx.WeakTypeTag](ctx: blackbox.Context)(script: ctx.Expr[WarpScript[T]]): ctx.Expr[WarpScript[T]] = {
+    // this untypecheck call is crucial to avoid crashing the compiler with mysterious errors like
+    // [Error] : Error while emitting MacrosSpec.scala
+    // value c
+    // https://github.com/scala/bug/issues/11628 is possibly related?
+    ctx.Expr[WarpScript[T]](testIdTransformer(ctx).transform(ctx.untypecheck(script.tree)))
+  }
+
+
+  /**
+    * Creates a Tree transformer that traverses a [[WarpScript]] AST and inserts test Id into `measure` calls.
+    *
+    * @param ctx blackbox macro Context.
+    * @return a (path-dependent) Tree transformer.
+    */
+  def testIdTransformer(ctx: blackbox.Context): ctx.universe.Transformer = {
+    import ctx.universe._
+
+    new Transformer {
+      val enclosingClass: String = extractEnclosingClass(ctx)
+      // TODO try to avoid using a var here
+      var capturedArg: Option[Tree] = None
+
+      override def transform(tree: Tree): Tree = tree match {
+        // capture arg here, but we dont need to transform
+        case Apply(func, args) if show(func).startsWith("com.workday.warp.monadic.WarpAlgebra.measure")
+            && (func.symbol.toString == "method map" || func.symbol.toString == "method flatMap") =>
+          capturedArg = args.headOption
+          super.transform(tree)
+
+        // a `measure` call. use the identifier we previously parsed to expand into a TestId
+        case Apply(func, args) if show(func).startsWith("com.workday.warp.monadic.WarpAlgebra.measure")
+            // TODO checking args length here is a suboptimal way to check whether a testId is already provided
+            && func.symbol.toString == "method measure" && args.length == 1 =>
+          val testId: String = s"""$enclosingClass.${extractSyntheticMethodName(ctx)(capturedArg.get)}"""
+          capturedArg = None
+          println(s"""macro expansion: inserting testId "$testId" into expression "$tree"""")
+          val newNode: Tree = Apply(func, Literal(Constant(testId)) :: args)
+          // quasiquote the transformed and parsed tree https://docs.scala-lang.org/overviews/quasiquotes/intro.html
+          q"${ctx.parse(show(newNode))}"
+
+        case _ =>
+          super.transform(tree)
+      }
+    }
+  }
+
+
+
+  /**
+    * Reifies the enclosing runtime class and returns the name of that class.
+    *
+    * @param ctx blackbox macro Context.
+    * @return fully qualified name of the enclosing runtime class.
+    */
+  @throws[MacroExpansionException]("when reifyEnclosingRuntimeClass does not return a Literal(Constant(_)) Tree node")
+  def extractEnclosingClass(ctx: blackbox.Context): String = {
+    import ctx.universe._
+    ctx.reifyEnclosingRuntimeClass match {
+      case Literal(Constant(className)) => className.toString
+      case other => throw new MacroExpansionException(s"""unable to extract class name from Tree node: ${showRaw(other)}""")
```

Comment:
What would case this case to be true?

Reply:
this case was true when i was trying to get this working for scala 2.11 and 2.12, they changed the shape of the tree returned by `reifyEnclosingRuntimeClass`. i left this case here just to avoid compiler warnings about non-exhaustive matching